### PR TITLE
=str fusing PR erratum

### PR DESCRIPTION
**Pull request by 2m**
Builds on top of #19104
- make tck tests compile
- make jmh tests compile
- remove assumptions from tests which do not hold anymore when fused
- add runBatch to OnSubscribe event, fixes tests of completely fused stages

### CODE REVIEW

File: akka-bench-jmh-dev/src/main/scala/akka/stream/FlowMapBenchmark.scala [hunk 1]
```diff
@@ -99,12 +96,7 @@ class FlowMapBenchmark {
       }
     }
 
-    flow = mkMaps(Source(syncTestPublisher), numberOfMapOps) {
-      if (UseGraphStageIdentity)
-        new GraphStages.Identity[Int]
-      else
-        Flow[Int].map(identity)
-    }
```
Comment (rkuhn) [inline akka-bench-jmh-dev/src/main/scala/akka/stream/FlowMapBenchmark.scala]: please keep this as a check of the effects of the emulation layer

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala [hunk 2]
```diff
@@ -37,8 +37,8 @@ object GraphStages {
    * INERNAL API
    */
   private[stream] abstract class SimpleLinearGraphStage[T] extends GraphStage[FlowShape[T, T]] {
-    val in = Inlet[T]("in")
-    val out = Outlet[T]("out")
+    val in = Inlet[T](toString + ".in")
+    val out = Outlet[T](toString + ".out")
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: good!

File: akka-bench-jmh-dev/src/main/scala/akka/stream/FlowMapBenchmark.scala (hunk 1 above)
Comment (2m) [inline akka-bench-jmh-dev/src/main/scala/akka/stream/FlowMapBenchmark.scala]: Right. Makes sense.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala (hunk 2 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: Doesn't this create a lot of retained memory?

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: And, AFAIK we never used the full string representation for these before. Shouldn't think be the getClass.simpleName + ".in"/".out"?

Comment (2m) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: Two string builders for every graph stage instance that inherits `SimpleLinearGraphStage`. Is there a more optimal  way to have properly named in(out)lets? Without this it is harder to follow debug output of fusing.

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: Does: `Logging.simpleName(this) + ".in"` and `Logging.simpleName(this) + ".out"` work?

Comment (2m) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: Yes, it does (`class TakeWithin[T](timeout: FiniteDuration) extends SimpleLinearGraphStage[T]`).

```
wiring MaybeSource.out (2145a4f4) -> IdleInject.in (0435d8d4)
wiring IdleInject.out (5d67b4c9) -> TakeWithin.in (7ed139c8)
wiring TakeWithin.out (2ec5b7f7) -> headOption.in (562fb04d)
```

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: That looks great! :+1:

On Tue, Dec 8, 2015 at 3:04 PM, Martynas Mickevičius <
notifications@github.com> wrote:

> In akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala
> <https://github.com/akka/akka/pull/19110#discussion_r46955700>:
>
> > @@ -37,8 +37,8 @@ object GraphStages {
> >     * INERNAL API
> >     */
> >    private[stream] abstract class SimpleLinearGraphStage[T] extends GraphStage[FlowShape[T, T]] {
> > -    val in = Inlet[T]("in")
> > -    val out = Outlet[T]("out")
> > +    val in = Inlet[T](toString + ".in")
> > +    val out = Outlet[T](toString + ".out")
>
> Yes, it does (class TakeWithin[T](timeout: FiniteDuration) extends
> SimpleLinearGraphStage[T]).
>
> wiring MaybeSource.out (2145a4f4) -> IdleInject.in (0435d8d4)
> wiring IdleInject.out (5d67b4c9) -> TakeWithin.in (7ed139c8)
> wiring TakeWithin.out (2ec5b7f7) -> headOption.in (562fb04d)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19110/files#r46955700>.
>



-- 
Cheers,
√

**Comment by 2m** [comment]
Removed delay from Merge prefetch and added a buffer for original FlowJoinSpec testcase.

**Comment by rkuhn** [comment]
Thanks, @2m!

**Pull request merged**